README.md
[![NPM](https://nodei.co/npm/yocto-router.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/yocto-router/)

![alt text](https://david-dm.org/yoctore/yocto-router.svg "Dependencies Status")
[![Code Climate](https://codeclimate.com/github/yoctore/yocto-router/badges/gpa.svg)](https://codeclimate.com/github/yoctore/yocto-router)
[![Test Coverage](https://codeclimate.com/github/yoctore/yocto-router/badges/coverage.svg)](https://codeclimate.com/github/yoctore/yocto-router/coverage)
[![Issue Count](https://codeclimate.com/github/yoctore/yocto-router/badges/issue_count.svg)](https://codeclimate.com/github/yoctore/yocto-router)
[![Build Status](https://travis-ci.org/yoctore/yocto-router.svg?branch=master)](https://travis-ci.org/yoctore/yocto-router)

## Overview

This module is a part of yocto node modules for NodeJS.

Please see [our NPM repository](https://www.npmjs.com/~yocto) for complete list of available tools (completed day after day).

This module manage routes configuration & mapping for an express app.

## Motivation

Create a simple and ready to use router manager from a json file configuration.

## How to use

```javascript
var logger    = require('yocto-logger');
var r         = require('yocto-router')(logger);
var express   = require('express');

var app = express();
// set path into app
r.useApp(app);
// set routes path
r.setRoutes('./example/routes/');
// set controller endpoint
r.setEndPoint('./example/ctrl/'); 
// process configure
if (!r.configure()) {
  // error
} else {
  // your valid process here
}
```

## Routes Json configuration files syntax

Your json configuration for routes must be follow these syntax : 

```javascript
[
  {
    "method"  : "YOUR_VALID_HTTP_METHOD_NAME_HERE",
    "path"    : "YOUR_ROUTE_PATTERN_HERE",
    "regexp"  : "true|false", // Eval string as a regexp or not
    "priority" : 100, // Defined your priority order on route definition into express
    "controller" : {
      "name"  : "YOUR_CONTROLLER_NAME_HERE",
      "fn"  : "YOUR_CONTROLLER_FN_NAME_HERE"
    }
  }
]
```
See Below an example : 

```javascript
[
  {
    "method"  : "post",
    "path"    : "/foo/bar",
    "controller" : {
      "name"  : "controllerFoo",
      "fn"  : "methodBar"
    }
  }  
]
```

## Routes for 404 and 500 HTTP codes

You can also defined route process for 404 and 500 HTTP code, just follow config example below : 

```javascript
[
  {
    "error" : 404 | 500 // with code you want
    "priority" : 100, // Defined your priority order on route definition into express very important
    "controller" : {
      "name"  : "YOUR_CONTROLLER_NAME_HERE",
      "fn"  : "YOUR_CONTROLLER_FN_NAME_HERE"
    }
  }
]
```

Into your controller just follow the express syntax for error middleware process.

You can find here the [Express FAQ for more details](http://expressjs.com/starter/faq.html)


## Controller Javascript files syntax

If you are here, it's because you have already define a least one route.

So controller syntax is a same than express export syntax.

```javascript
exports.methodBar = function(req, res, next) {
  // your code here
};
```

## Context in your controller enpoint

In your controller endpoint you can use `this` keywords to access on your express app

## Logging in tool

By Default this module include [yocto-logger](https://www.npmjs.com/package/yocto-logger) for logging.
It's possible to inject in your router instance your current logger instance if is another `yocto-logger` instance.

For example : 

```javascript 
var logger = require('yocto-logger');
// EXTRA CODE HERE FOR YOUR APP
// AGAIN & AGAIN
var router = require('yocto-router')(logger);
```

## Changelog

All history is [here](https://github.com/yoctore/yocto-router/blob/master/CHANGELOG.md)

CHANGELOG.md
## 2.0.1 (2018-02-22)

- Bug fix when bind function

## 2.0.0 (2018-02-16)

- Major update of all dependencies
- Update node version engine to >= 8.9.4

## 1.3.4 (2016-03-31)

- Update package json
- Add unit tests
- Update README.md

## 1.3.2 (2015-11-18)

- Now we return true when no route was founded. we display just a warning.

In some case this is very useful, to not stop app.

## 1.3.1 (2015-11-13)

- Update package dependencies

## 1.3.0 (2015-11-12)

- Add 404 & 500 process on route config

## 1.2.0 (2015-11-12)

- Add priority property on route definition to defined route ordre properly

## 1.1.0 (2015-11-05)

- Add Regexp process on route matching

## 1.0.3 (2015-11-03)

- Bind app on endpoint function

## 1.0.1 (2015-10-23)

- Change a logger message

## 1.0.0

- Add core process
- Add utility method to set routes and controller
- Change js file to json file for routes.
- Fix Minor bugs
- Add Readme.md content
- Prepare to publish


test/1-http-test.js
'use strict';

var chai      = require('chai').assert;
var expect    = require('chai').expect;
var should    = require('chai').should();
var logger    = require('yocto-logger');
var r         = require('../src')(logger);
var express   = require('express');
var _         = require('lodash');
var path      = require('path');

// logger.disableConsole();

// routes list
var routes = [
  { name : 'routes', path : './example/routes/', endpoint : false },
  { name : 'ctrl', path : './example/ctrl/', endpoint : true }
];

var app = express();

describe('Render ->', function () {
  describe('Set and Get process must be valid', function () {
    it ('Set params to render must be valid', function () {
      app.set('render', { foo : 'bar' });
      // get value
      var r = app.get('render');
      // test case
      expect(r).to.not.empty;
      expect(r).to.be.an('object');
      expect(r).to.deep.equal({ foo : 'bar' });
    });

    it ('Attached router with app must be succeed', function () {
      var state = r.useApp(app);
      // unit test
      expect(state).to.be.a('boolean');
      expect(state).to.equal(true);
    });
  });

  // parse each routes
  _.each(routes, function (ro) {
    describe([ 'Loading config for ->', ro.name, ].join(' '), function () {
      it ([ 'Except be succeed when loading', ro.path ].join(' '), function () {
        // default state
        var state = false;
        // has endpoint
        if (ro.endpoint) {
          state = r.setEndPoint(ro.path);
        } else {
          state = r.setRoutes(ro.path);
        }
        // unit test
        expect(state).to.be.a('boolean');
        expect(state).to.equal(true);
      });
      it ([ 'Except setted path to be the strict equal to ', ro.path ].join(' '), function () {
        var fpath = path.normalize([ process.cwd(), ro.path ].join('/'));
        expect(fpath).to.be.a.string;
        expect(fpath).to.equal(r[ro.name]);
      });
    });
  });

  describe('Configure process must be succeed', function () {
    it ('Expect to be a boolean and equal to true', function () {
      var state = r.configure();
      // unit test
      expect(state).to.be.a('boolean');
      expect(state).to.equal(true);
    });
    it ('Expect to be a not empty string and a type of [ GET | POST | PUT | DELETE | PATCH ] route',
      function () {
      app._router.stack.forEach(function(r) {
        if (r.route && r.route.path) {
          expect(r.route.path).to.be.not.empty;
          expect(r.route.path).to.be.a.string;
          expect([ 'get', 'post', 'put', 'patch', 'delete' ]).to.be.include.members(Object.keys(r.route.methods));
        }
      });
    });
  });
});


example/routes/test.js
/**
 * Default routes for ad request
 */
var routes = [
  {
    method  : 'post',
    path    : '/ad/contact/details/:property?',
    controller : {
      name  : 'ad',
      fn  : 'fnB'
    }
  },
  {
    method  : 'post',
    path    : '/ad/toto',
    controller : {
      name  : 'ad',
      fn  : 'fnA'
    }
  }  
];

/**
 * expose routes
 */
exports.routes = routes;




Gruntfile.js
'use strict';

module.exports = function (grunt) {
  // Init config
  grunt.initConfig({
    // Default package
    pkg : grunt.file.readJSON('package.json'),

    // Hint our app
    yoctohint : {
      json : [
        'package.json'
      ],
      node : [
        'src/*.js', 'Gruntfile.js'
      ],
      options : {
        compatibility : true
      }
    },

    // Uglify our app
    uglify : {
      options : {
        banner : '/* <%= pkg.name %> - <%= pkg.description %> - V<%= pkg.version %> */\n'
      },
      api : {
        files : [ {
          expand : true,
          cwd    : 'src',
          src    : '**/*.js',
          dest   : 'dist'
        } ]
      }
    },

    // Unit tests
    mochacli : {
      options : {
        reporter       : 'spec',
        'inline-diffs' : false,
        'no-exit'      : true,
        force          : false,
        'check-leaks'  : true,
        bail           : false
      },
      all : [ 'test/*.js' ]
    }
  });

  // Load the plugins
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-mocha-cli');
  grunt.loadNpmTasks('yocto-hint');

  grunt.registerTask('hint', 'yoctohint');
  grunt.registerTask('test', 'mochacli');
  grunt.registerTask('build', [ 'hint', 'test', 'uglify' ]);
  grunt.registerTask('default', 'build');
};


example/ctrl/ad.js
exports.notFound = function(req, res) {
  // to do fn A
  console.log('dans not found');
  res.status(400).send('Sorry cant find that!');
};

exports.systemError = function(err, req, res, next) {
  // to do fn A
  console.log('dans systemError');
  res.status(500).send('Sorry cant find that!');
};


exports.other = function(req, res) {
  // to do fn A
  console.log('dans other');
  res.send(200).end();
};

exports.other2 = function(req, res) {
  // to do fn A
  console.log('dans other2');
  res.send(200).end();
};

exports.end = function(req, res) {
  // to do fn A
  console.log('dans end');
  res.send(200).end();
};

exports.test1 = function(req, res) {
  // to do fn A
  console.log('dans test1');
  res.send(200).end();
};

exports.test2 = function(req, res) {
  // to do fn A
  console.log('dans test2');
  res.send(200).end();
};

src/index.js
'use strict';

var logger        = require('yocto-logger');
var glob          = require('glob');
var _             = require('lodash');
var path          = require('path');
var joi           = require('joi');
var utils         = require('yocto-utils');
var fs            = require('fs');

/**
 * Manage route for node js app
 *
 * @class Router
 * @param {Object} logger Logger instance
 */
function Router (logger) {
  /**
   * Default logger instance
   *
   * @property logger
   * @type Object
   */
  this.logger = logger;

  /**
   * Default base path for routes
   *
   * @property routes
   * @type String
   */
  this.routes = process.cwd();

  /**
   * Default base path for controller assign on routes
   *
   * @property ctrl
   * @type String
   */
  this.ctrl = process.cwd();

  /**
   * Default app value must be an express app
   *
   * @property app
   * @type String
   * @default Null
   */
  this.app = null;
}

/**
 * Default configure method, use for main process of module
 *
 * @method configure
 * @return {Boolean} true if all is ok false otherwise
 */
Router.prototype.configure = function () {
  // Default list of routes
  var routesList = [];

  // Base message
  this.logger.banner('[ Router.configure ] - Initializing Router ...');

  // Main process
  try {
    // Check app
    if (!_.has(this.app, 'use') || !_.isFunction(this.app.use)) {
      throw 'Your app is invalid. Please ad a valid express app on your Router.app property';
    }

    // Getting toutes
    var routes = glob.sync('*.json', {
      cwd      : this.routes,
      nosort   : true,
      realpath : true
    });

    // Empty routes ?
    if (_.isEmpty(routes)) {
      this.logger.warning([ '[ Router.configure ] - No routes founded on',
        this.routes ].join(' '));
    }

    // Routes
    _.each(routes, function (item) {
      // Get controller name
      var ctrlName = item.replace(this.routes, '');

      // Logging message
      this.logger.info([ '[ Router.configure ] - Retreiving routes for [',
        ctrlName, ']' ].join(' '));

      // Require module
      var mods = JSON.parse(fs.readFileSync(item));

      // Has routes ?
      if (_.isEmpty(mods)) {
        // War message
        this.logger.warning([ '[ Router.configure ] - Module [', ctrlName,
          '] doesn\'t have any routes.',
          'Remove this file or add one route configuration' ].join(' '));
      } else {
        // Parses routes
        _.each(mods, function (mod) {
          // Validation schema
          var schema = joi.object().min(3).max(5).keys({
            method : joi.string().required().empty().valid([
              'get', 'post', 'put', 'delete', 'options', 'head'
            ]),
            path       : joi.string().required().empty().min(1),
            regexp     : joi.boolean().default(false),
            priority   : joi.number().default(100).min(0).max(999),
            controller : joi.object().required().min(2).max(2).keys({
              name : joi.string().required().empty().min(1),
              fn   : joi.string().required().empty().min(1)
            }).allow('name', 'fn')
          }).allow('method', 'path', 'controller', 'regexp', 'priority');

          // Is error path ? so need to clean validation rules
          if (_.has(mod, 'error')) {
            // Defined error schema
            schema = joi.object().min(3).max(3).keys({
              priority   : joi.number().default(100).min(0).max(999),
              error      : joi.number().optional().allow([ 404, 500 ]),
              controller : joi.object().required().min(2).max(2).keys({
                name : joi.string().required().empty().min(1),
                fn   : joi.string().required().empty().min(1)
              }).allow('name', 'fn')
            }).allow('controller', 'priority', 'error');
          }

          // Validate
          var result = schema.validate(mod);

          // Has error
          if (!_.isNull(result.error)) {
            // Parse details
            _.each(result.error.details, function (error) {
              this.logger.warning([ '[ Router.configure ] - Cannot add route for',
                ctrlName,
                'config given is invalid. Error is :',
                utils.obj.inspect(error)
              ].join(' '));
            }.bind(this));
          } else {
            // Override mod with joi validate value
            mod = result.value;

            // Build ctrlPath
            var ctrlPath =  path.normalize([ [ this.ctrl, mod.controller.name ].join('/'), 'js'
            ].join('.'));

            // File exits ?
            if (!fs.existsSync(ctrlPath)) {
              // It looks like no ....
              this.logger.warning([ '[ Router.configure ] - Given endpoint controller [',
                mod.controller.name, '] for route [', ctrlName,
                '] is invalid. Path [', ctrlPath,
                '] is not found. Operation Aborted !' ].join(' '));
            } else {
              // Require controller
              var controller = require(ctrlPath);

              // Func exists and is a func ??
              if (!_.has(controller, mod.controller.fn) ||
                  !_.isFunction(controller[mod.controller.fn])) {
                // Warning
                this.logger.warning([ '[ Router.configure ] - Cannot find Function [',
                  mod.controller.fn, '] for controller [',
                  mod.controller.name, '] on [', ctrlPath,
                  ']. Operation Aborted' ].join(' '));
              } else {
                // Messsage
                this.logger.info([ '[ Router.configure ] - Adding route [',
                  mod.path || mod.error, '] on a [',
                  mod.method ? mod.method.toUpperCase() : mod.error,
                  '] HTTP Request with a callback on [',
                  [ mod.controller.name, mod.controller.fn ].join('.'),
                  ']' ].join(' '));

                // Has regexp
                if (_.has(mod, 'regexp') && mod.regexp) {
                  // Path to regexp
                  mod.path = new RegExp(mod.path);
                }

                // Save priority
                var priority = mod.priority;

                // Remove key

                delete mod.priority;

                // Push routes item
                routesList.push({
                  item       : mod,
                  controller : controller,
                  priority   : priority
                });
              }
            }
          }
        }.bind(this));
      }
    }.bind(this));
  } catch (e) {
    // Something is invalid
    this.logger.error([ '[ Router.configure ] - An Error occured during routes initialization.',
      'error is :', e, 'Operation aborted !' ].join(' '));

    // Statement
    return false;
  }

  // Return true if all is valid
  return this.addRoute(routesList);
};

/*
 * Default method to add un route on current app. This method will sort route by given priority.
 * By default this priority was defined to 100
 *
 * @param {Array} routes default route list
 * @return {Boolean} return true if all is ok false otherwise
 */
Router.prototype.addRoute = function (routes) {
  // Has routes ?
  if (routes.length > 0) {
    // Process routes
    routes = _.sortBy(routes, function (route) {
      // Default sort statement
      return this.min(route.priority);
    }.bind(Math));

    // Parse all routes
    _.each(routes, function (route) {
      // Is normal route ?
      if (!_.has(route.item, 'error')) {
        // Route
        this.app[route.item.method](route.item.path,
          route.controller[route.item.controller.fn].bind(this.app));
      } else {
        // Default handler for error
        this.app.use(route.controller[route.item.controller.fn].bind(this.app));
      }
    }.bind(this));

    // How many routes added ??
    this.logger.info([ '[ Router.addRoute ] -', routes.length,
      routes.length < 2 ? 'route' : 'routes',
      'was added on current app' ].join(' '));
  } else {
    // Log message
    this.logger.warning('[ Router.addRoute ] - No routes to add.');
  }

  // Default statement
  return true;
};

/**
 * Set path for routes files
 *
 * @param {String} path path of js files
 * @return {Boolean} true if all is ok false otherwise
 */
Router.prototype.setRoutes = function (path) {
  // Default statement
  return this.set('routes', path);
};

/**
 * Set path for controller files
 *
 * @param {String} path path of js files
 * @return {Boolean} true if all is ok false otherwise
 */
Router.prototype.setEndPoint = function (path) {
  // Default statement
  return this.set('ctrl', path);
};

/**
 * Define app reference for mapping
 *
 * @param {Object} app express app to use
 * @return {Boolean} true if all is ok false otherwise
 */
Router.prototype.useApp = function (app) {
  // Default statement
  return this.set('app', app);
};

/**
 * Set function, assign given data to a property
 *
 * @param {String} name name of property
 * @param {String} value value of property
 * @return {Boolean} true if all is ok false otherwise
 */
Router.prototype.set = function (name, value) {
  // Check requirements
  if (!_.isUndefined(name) && _.isString(name) && !_.isEmpty(name)) {
    // Is relative ?
    if ((name === 'routes' || name === 'ctrl') && !path.isAbsolute(value)) {
      // Process correct path
      value = path.normalize([ process.cwd(), value ].join('/'));

      // Try here exception can be throwed
      try {
        // Parse file
        var parse   = fs.statSync(value);

        // Change state

        if (!parse.isDirectory()) {
          // Throw exception here
          throw [ 'Path is not a directory for :', name ].join(' ');
        }
      } catch (e) {
        // Warn message
        this.logger.warning([ '[ Router.set ] - given value is not a directory : ', e ].join(' '));

        // Return invalid statement if we are here
        return false;
      }
    }

    // Assign value
    this[name] = value;
  } else {
    // Log a warning messsage.
    this.logger.warning([ '[ Router.set ] - Invalid value given.',
      'name must be a string and not empty. Operation aborted !' ].join(' '));

    // Invalid statement
    return false;
  }

  // Returning current instance
  return true;
};

// Default export
module.exports = function (l) {
  // Is a valid logger ?
  if (_.isUndefined(l) || _.isNull(l)) {
    logger.warning('[ Router.constructor ] - Invalid logger given. Use internal logger');

    // Assign
    l = logger;
  }

  // Default statement
  return new Router(l);
};
